record/envelope.go

```diff
@@ -0,0 +1,196 @@
+package record
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"time"
+
+	pool "github.com/libp2p/go-buffer-pool"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	pb "github.com/libp2p/go-libp2p-core/record/pb"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/multiformats/go-varint"
+)
+
+// SignedEnvelope contains an arbitrary []byte payload, signed by a libp2p peer.
+//
+// Envelopes are signed in the context of a particular "domain", which is a
+// string specified when creating and verifying the envelope. You must know the
+// domain string used to produce the envelope in order to verify the signature
+// and access the payload.
+type SignedEnvelope struct {
+	// The public key that can be used to verify the signature and derive the peer id of the signer.
+	PublicKey crypto.PubKey
+
+	// A binary identifier that indicates what kind of data is contained in the payload.
+	// TODO(yusef): enforce multicodec prefix
+	PayloadType []byte
+
+	// A monotonically-increasing sequence counter for ordering SignedEnvelopes in time.
+	Seq uint64
+
+	// The envelope payload.
+	Payload []byte
+
+	// The signature of the domain string :: type hint :: payload.
+	signature []byte
+}
```

Comment:
Can we add a `cached record.Record` field where we stash the unmarshalled payload? We should also add an accessor `Record() Record`. Rationale: we might want to parse and validate an envelope in some layer, then pass the envelope to another layer (e.g. peerstore), omitting the cost of re-unmarshalling.

Reply:
I like this, but if there are cases where we construct an `Envelope` without unmarshalling the payload, the `Record` accessor would need to fallback to unmarsahlling if the cached record doesn't exist. In that case, it seems like it should return `(Record, error)` in case unmarshalling fails.

If we did that and had the accessor serve as a "lazy-loading getter", we could potentially skip unmarshalling the record in `ConsumeEnvelope`, and have the unmarshalling always happen in the first call of the accessor function. We could also add a `TypedRecord(dest Record) error` accessor to use a specific `Record` type.

That would let us validate the signature of envelopes whose payloads we don't know how to parse. I'm not sure why/if we want to do that; the only use case I can think of is storing and forwarding `Envelope`s for third-party peers who may know about different payload types than we do.